doc: Add documentation on transpilation (#681)
SAP/cloud-sdk: docs/js/features/odata/generate-odata-client.mdx

  - generate
---

With the generator you can generate custom OData client libraries for OData services.
You can then access these services using the client libraries.

> You can use the SAP Cloud SDK generator both as a command-line interface (CLI) and programmatically.

Author: I removed the passive voice, that's why ;)
Now, I changed it a bit, let me know what you think.

All you need to use it, is a service metadata specification in the `EDMX` format (file ending can be `.edmx` or `.xml`).

## Installation

Run this command in your project's terminal to install the generator as a `devDependency`:

```sh
npm install -D @sap-cloud-sdk/generator
```

or globally:

```sh
npm install -g @sap-cloud-sdk/generator
```

:::note

> If you want to generate JavaScript code and install the generator globally, you have to install the `@sap-cloud-sdk/odata-v2` or `@sap-cloud-sdk/odata-v4` package in your project.

Reviewer: This seems helpful 👍 I like it

The generation always includes the TypeScript sources.
All other files can be excluded from the generation - see the options below.

## Transpilation

The SAP Cloud SDK OData client generator for JavaScript/TypeScript generates TypeScript code.

> By default, it transpiles the generated sources to JavaScript and provides a `package.json` for the client (this differs from the OpanAPI generator).

Reviewer: ```suggestion
By default, it transpiles the generated sources to JavaScript and provides a `package.json` for the client (this differs from the OpenAPI generator).
```
As a reader I ask myself: Why does it provide a `package.json`? Especially if I need to install dependencies manually anyways...

> The generated clients depend on the `@sap-cloud-sdk/odata-v2` or `@sap-cloud-sdk/odata-v4` packages, depending on the OData version of the service.

Reviewer: ```suggestion
The generated clients depend on the `@sap-cloud-sdk/odata-v2` or `@sap-cloud-sdk/odata-v4` packages, depending on the OData version of the service.
```
As a reader: Is this reflected in the generated package.json?

> Therefore, for transpilation you need a local reference to those packages.

Reviewer: ```suggestion
Therefore, you need a local reference to those packages for transpilation.
```

> Those packages are not installed for you, so that you can decide where to install them (e.g. in your project or in the subdirectory of your client).

Reviewer: ```suggestion
Those packages are not installed automatically. You need to install them either in your project or in the client directory.
```
Installing in the client directory is (AFAIK) not possible without at least running the build once before (either failing it during transpilation or skipping transpilation via flag). Then you would need to run the generator again but make sure to not "cleanOutputDir"... otherwise the dependency is gone again and it fails...

> If you installed the generator globally, transpilation will fail unless you install the required odata packages.

Reviewer: ```suggestion
If you installed the generator globally, transpilation will fail unless you install the required package (`sap-cloud-sdk/odata-v2` or `sap-cloud-sdk/odata-v4`).
```

> If you installed the generator as a `devDependency`, transpilation will work, however, for productive use, you may have to define the dependency productively.

Reviewer: I am confused... if this is the case why would I install it as a devDependency in the first place? But why would this be the case? Why "may have to"? What is different in production?

Reviewer: ```suggestion
If you installed the generator as a `devDependency`, transpilation will work automatically.
However, for productive use, you may have to define the dependency productively.
```
